From 39eb2ad22178acb6dc43de991688cbaa6ecf50d0 Mon Sep 17 00:00:00 2001
From: Takuro Ashie <ashie@homa.ne.jp>
Date: Mon, 10 Apr 2017 17:03:35 +0900
Subject: [PATCH 35/44] GLLibraryEGL: Use wl_display to get EGLDisplay on
 Wayland

Because some drivers doesn't support EGL_DEFAULT_DISPLAY.

Signed-off-by: Hiroshi Hatake <hatake@clear-code.com>

This patch is partially imported from
109766aa316cc3b2261bd7eac81efc823af1da82.
---
 gfx/gl/GLLibraryEGL.cpp | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/gfx/gl/GLLibraryEGL.cpp b/gfx/gl/GLLibraryEGL.cpp
index 130bce119152..105fc28f68fb 100644
--- a/gfx/gl/GLLibraryEGL.cpp
+++ b/gfx/gl/GLLibraryEGL.cpp
@@ -27,6 +27,12 @@
 #include "GLContextProvider.h"
 #include "gfxPrefs.h"
 #include "ScopedGLHelpers.h"
+#ifdef MOZ_WIDGET_GTK
+#include <gdk/gdk.h>
+#ifdef GDK_WINDOWING_WAYLAND
+#include <gdk/gdkwayland.h>
+#endif // MOZ_WIDGET_GTK
+#endif // GDK_WINDOWING_WAYLAND
 
 namespace mozilla {
 namespace gl {
@@ -510,7 +516,19 @@ GLLibraryEGL::EnsureInitialized(bool forceAccel, nsACString* const out_failureId
             mIsWARP = true;
         }
     } else {
-        chosenDisplay = GetAndInitDisplay(*this, EGL_DEFAULT_DISPLAY);
+        void *nativeDisplay = EGL_DEFAULT_DISPLAY;
+#ifdef GDK_WINDOWING_WAYLAND
+        // Some drivers doesn't support EGL_DEFAULT_DISPLAY
+        GdkDisplay *gdkDisplay = gdk_display_get_default();
+        if (GDK_IS_WAYLAND_DISPLAY(gdkDisplay)) {
+            nativeDisplay = gdk_wayland_display_get_wl_display(gdkDisplay);
+            if (!nativeDisplay) {
+              NS_WARNING("Failed to get wl_display.");
+              return false;
+            }
+        }
+#endif
+        chosenDisplay = GetAndInitDisplay(*this, nativeDisplay);
     }
 
     if (!chosenDisplay) {
-- 
2.11.0

